lib/phoenix/pubsub/local.ex

```diff
@@ -129,39 +173,90 @@ defmodule Phoenix.PubSub.Local do
 
   @doc false
   # This is an expensive and private operation. DO NOT USE IT IN PROD.
-  def list(local_server) when is_atom(local_server) do
-    local_server
+  def list(pubsub_server, shard) when is_atom(pubsub_server) do
+    shard
+    |> local_for_shard(pubsub_server)
     |> :ets.select([{{:'$1', :_}, [], [:'$1']}])
     |> Enum.uniq
   end
 
   @doc false
-  # This is an expensive and private operation. DO NOT USE IT IN PROD.
-  def subscription(local_server, pid) when is_atom(local_server) do
-    local_server
-    |> :ets.select([{{:'$1', {pid, :_}}, [], [:'$1']}])
-    |> Enum.uniq
+  def subscription(pubsub_server, pool_size, pid) when is_atom(pubsub_server) do
+    pid
+    |> pid_to_shard(pool_size)
+    |> local_for_shard(pubsub_server)
+    |> GenServer.call({:subscription, pid})
+  end
+
+  @doc false
+  def local_name(pubsub_server, shard) do
```

Comment:
Nitpick: I don't like those two functions. I think those names should be generated just in the supervisor init and adding those functions will allow us to be lazy in the future and start relying on them at runtime.